scripts/get-workspace-map.sh

```diff
@@ -0,0 +1,30 @@
+#!/bin/bash
+# get-workspace-map.sh context-1 context-2... context-n
+#   Read Namespace resources and generate a JSON object mapping from namespaces
+#   to workspace based on the value of label toolchain.dev.openshift.com/space.
+#   This script assumes `oc` is preconfigured with all the required clusters
+#   and could be found in $PATH
+#
+set -o pipefail -o errexit -o nounset
+
+out={}
+for context in "$@"
```

Comment:
Id prefer we use an env var with some useful default value rather than requiring parameters to be explicitly set in each call

Reply:
What might be a good default value for that?

Reply:
What I'd really love to have happen here is that if you clone the repo and run the script locally it just works - assuming you already have permissions to the clusters somehow. How close can we get to that? Can we somehow make is so the context definitions come from some file in the repo while the connection credentials come from the user's `~/.kube/config` file? As an alternative, perhaps we can leverage openid?

Reply:
What if we assume that the context is defined and we have the script grep for lines that has the cluster name in it?

e.g. if we run it for stage, we use the user's pre-stored contexts that include `stg-m01` or `stg-rh01`. This way, we don't have to know the exact name on one hand (for me it's stored with a name that includes my github username), and we don't have to know how to fetch it on the other hand.

Reply:
Given that `.kube/config` is a YAML file - `grep` would not be the right tool for this, `yq` would, but that's adding another dependency. Maybe if we grep on `oc config get-contexts`.

Ultimately this feels like a bit of an overkill that is not reusable for the "running as an automated job" use case. I would prefer to have a pre-backed kubeconfig file to neutralize any garbage that may be in the user's file.

I really feel like OpenID with the `oidc-login` plugin may work - are you familiar with how that is used to connect to our systems?

Reply:
(I didn't mean using the grep toll necessarily - just to rely on the contexts that the user has)

I'm not familiar with anything related to OpenID other than the fact that I click on it as authentication option from time to time.

Reply:
I feel we have two different tasks here that make each other more complicated.

What if we make the script to accept the contexts as it is at the moment, and we later on have it run another script to fetch the contexts in case contexts are not provided?

We can then implement the process of fetching the contexts independently of the current issue, and that process can also be used for other scripts to require something similar.

Reply:
Have a look here: 
https://docs.google.com/document/d/1hFvQDH1H6MGNqTGfcZpyl2h8OIaynP8sokZohCS0Su0/edit#heading=h.olhho0rpp8t5

This document actually complicates things because of the need to use the RHTAP proxy. Using the `kubectl` [oidc-login][1] plug-in is even simpler. Once you have things set up you get to the point where using a properly configured context just works, or otherwise pops up your browser to login to the cluster if you're logged out.

If you've never installed a kubectl/oc plug-in before (Any kubectl plug-in also works with oc), I recommend using [krew][2] for this. You can obtain krew itself [via asdf][3] if you're using it.

[1]: https://github.com/int128/kubelogin
[2]: https://krew.sigs.k8s.io/
[3]: https://github.com/nlamirault/asdf-krew

Reply:
I don't want to defer the problem of figuring out which contexts to use - its an inherent part of writing this script (Regardless of wither you factor it out to a different script or not). I don't foresee other scripts having to solve a similar problem for now.

Reply:
So a quick attempt at setting things up via OIDC failed. I added this context to my configuration:
```
- context:
    cluster: api-stone-stg-m01-7ayg-p1-openshiftapps-com:6443
    namespace: default
    user: oidc
  name: rhtap-stage-m01
```

But trying to use that gave me: 
```
error: You must be logged in to the server (Unauthorized)
```

This is despite something very similar working well via the RHTAP proxy (Effectively going to the same cluster).

Here is the setup for the oidc user, the `rhoas-cli-prod` thing makes me suspect there is more here then just going via `sso.redhat.com` which all clusters do.
```
- name: oidc
  user:
    exec:
      apiVersion: client.authentication.k8s.io/v1beta1
      args:
      - oidc-login
      - get-token
      - --oidc-issuer-url=https://sso.redhat.com/auth/realms/redhat-external
      - --oidc-client-id=rhoas-cli-prod
      command: oc
      env: null
      interactiveMode: IfAvailable
      provideClusterInfo: false
```

So I suppose its back to grepping the user's existing configuration.

Reply:
Any way we turn it, I think one thing is missing from the script at the moment is being able to handle failures. As I see it, the script needs to replace the output from clusters it failed fetching with `{}` and to somehow report that at the end.

Reply:
I think we might get that for free, if `oc` fails to connect it would dump some error into STDERR and nothing to STDOUT, so we'll end up with `{}` from `jq` - but you probably should test that assumption.

Reply:
This is done now. Probably there are some ways to make it prettier, but let me know if you're ok with the approach.

Reply:
@ifireball This is done now. Probably there are some ways to make it prettier, but let me know if you're ok with the approach.